File: server/src/models/Campaign.model.js
const { DataTypes } = require('sequelize');
const { sequelize } = require('../config/database');

const Campaign = sequelize.define(
  'Campaign',
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
      allowNull: false,
    },
    advertiserId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'advertisers',
        key: 'id',
      },
      onDelete: 'CASCADE',
      field: 'advertiser_id',
    },
    name: {
      type: DataTypes.STRING(100),
      allowNull: false,
    },
    objective: {
      type: DataTypes.ENUM(
        'awareness',
        'traffic',
        'engagement',
        'app_installs',
        'conversions',
        'video_views'
      ),
      allowNull: false,
    },
    status: {
      type: DataTypes.ENUM(
        'draft',
        'pending_review',
        'active',
        'paused',
        'completed',
        'rejected'
      ),
      defaultValue: 'draft',
    },
    // Budget
    budgetType: {
      type: DataTypes.ENUM('daily', 'lifetime'),
      defaultValue: 'daily',
      field: 'budget_type',
    },
    budgetAmount: {
      type: DataTypes.INTEGER, // in cents
      allowNull: false,
      validate: {
        min: 100, // minimum $1.00
      },
      field: 'budget_amount',
    },
    budgetSpent: {
      type: DataTypes.INTEGER, // in cents
      defaultValue: 0,
      field: 'budget_spent',
    },
    // Bid strategy
    bidStrategy: {
      type: DataTypes.ENUM('lowest_cost', 'cost_cap', 'bid_cap'),
      defaultValue: 'lowest_cost',
      field: 'bid_strategy',
    },
    bidAmount: {
      type: DataTypes.INTEGER, // in cents
      allowNull: true,
      field: 'bid_amount',
    },
    // Schedule
    startDate: {
      type: DataTypes.DATE,
      allowNull: false,
      field: 'start_date',
    },
    endDate: {
      type: DataTypes.DATE,
      allowNull: true,
      field: 'end_date',
    },
    timezone: {
      type: DataTypes.STRING(50),
      defaultValue: 'UTC',
    },
    activeDays: {
      type: DataTypes.JSONB,
      defaultValue: [0, 1, 2, 3, 4, 5, 6],
      field: 'active_days',
    },
    activeHoursStart: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'active_hours_start',
    },
    activeHoursEnd: {
      type: DataTypes.INTEGER,
      defaultValue: 23,
      field: 'active_hours_end',
    },
    // Targeting
    ageMin: {
      type: DataTypes.INTEGER,
      defaultValue: 18,
      field: 'age_min',
    },
    ageMax: {
      type: DataTypes.INTEGER,
      defaultValue: 65,
      field: 'age_max',
    },
    gender: {
      type: DataTypes.ENUM('all', 'male', 'female'),
      defaultValue: 'all',
    },
    locations: {
      type: DataTypes.JSONB,
      defaultValue: [], // list of {country, state, city}
    },
    interests: {
      type: DataTypes.JSONB,
      defaultValue: [],
    },
    devices: {
      type: DataTypes.JSONB,
      defaultValue: ['all'],
    },
    includeCustomAudience: {
      type: DataTypes.JSONB,
      defaultValue: [],
      field: 'include_custom_audience',
    },
    excludeCustomAudience: {
      type: DataTypes.JSONB,
      defaultValue: [],
      field: 'exclude_custom_audience',
    },
    retargetEngaged: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: 'retarget_engaged',
    },
    retargetVisitors: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: 'retarget_visitors',
    },
    // Placements
    placementFeed: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      field: 'placement_feed',
    },
    placementReels: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      field: 'placement_reels',
    },
    placementStories: {
      type: DataTypes.BOOLEAN,
      defaultValue: true,
      field: 'placement_stories',
    },
    placementExplore: {
      type: DataTypes.BOOLEAN,
      defaultValue: false,
      field: 'placement_explore',
    },
    // Cumulative Analytics Metrics
    impressions: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    clicks: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    skips: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    videoViews: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'video_views',
    },
    videoCompletions: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'video_completions',
    },
    reach: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    frequency: {
      type: DataTypes.FLOAT,
      defaultValue: 0,
    },
    websiteClicks: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'website_clicks',
    },
    appInstalls: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
      field: 'app_installs',
    },
    purchases: {
      type: DataTypes.INTEGER,
      defaultValue: 0,
    },
    // Moderation
    rejectionReason: {
      type: DataTypes.STRING(255),
      allowNull: true,
      field: 'rejection_reason',
    },
    reviewedAt: {
      type: DataTypes.DATE,
      allowNull: true,
      field: 'reviewed_at',
    },
    reviewedBy: {
      type: DataTypes.UUID,
      allowNull: true,
      references: {
        model: 'users',
        key: 'id',
      },
      onDelete: 'SET NULL',
      field: 'reviewed_by',
    },
    // VIRTUAL COMPUTED GETTERS
    ctr: {
      type: DataTypes.VIRTUAL,
      get() {
        const imps = this.getDataValue('impressions') || 0;
        const clks = this.getDataValue('clicks') || 0;
        if (imps === 0) return '0.00';
        return ((clks / imps) * 100).toFixed(2);
      },
    },
    cpm: {
      type: DataTypes.VIRTUAL,
      get() {
        const imps = this.getDataValue('impressions') || 0;
        const spent = this.getDataValue('budgetSpent') || 0;
        if (imps === 0) return '0.00';
        return ((spent / imps) * 1000).toFixed(2);
      },
    },
    cpc: {
      type: DataTypes.VIRTUAL,
      get() {
        const clks = this.getDataValue('clicks') || 0;
        const spent = this.getDataValue('budgetSpent') || 0;
        if (clks === 0) return '0.00';
        return (spent / clks).toFixed(2);
      },
    },
  },
  {
    tableName: 'campaigns',
    timestamps: true,
    underscored: true,
  }
);

module.exports = Campaign;

File: client/lib/features/ads/data/models/campaign_model.dart
class CampaignModel {
  final String id;
  final String advertiserId;
  final String name;
  final String objective;
  final String status;
  final String budgetType;
  final int budgetAmount;
  final int budgetSpent;
  final String bidStrategy;
  final int? bidAmount;
  final DateTime startDate;
  final DateTime? endDate;
  final String timezone;
  final List<int> activeDays;
  final int activeHoursStart;
  final int activeHoursEnd;
  final Map<String, dynamic> targeting;
  final bool placementFeed;
  final bool placementReels;
  final bool placementStories;
  final bool placementExplore;

  // Analytics Metrics
  final int impressions;
  final int clicks;
  final int skips;
  final int videoViews;
  final int videoCompletions;
  final int reach;
  final double frequency;
  final int websiteClicks;
  final int appInstalls;
  final int purchases;

  // Review status
  final String? rejectionReason;
  final DateTime? reviewedAt;
  final String? reviewedBy;

  // Business Name from populating (admin route)
  final String? businessName;

  CampaignModel({
    required this.id,
    required this.advertiserId,
    required this.name,
    required this.objective,
    required this.status,
    required this.budgetType,
    required this.budgetAmount,
    required this.budgetSpent,
    required this.bidStrategy,
    this.bidAmount,
    required this.startDate,
    this.endDate,
    required this.timezone,
    required this.activeDays,
    required this.activeHoursStart,
    required this.activeHoursEnd,
    required this.targeting,
    required this.placementFeed,
    required this.placementReels,
    required this.placementStories,
    required this.placementExplore,
    required this.impressions,
    required this.clicks,
    required this.skips,
    required this.videoViews,
    required this.videoCompletions,
    required this.reach,
    required this.frequency,
    required this.websiteClicks,
    required this.appInstalls,
    required this.purchases,
    this.rejectionReason,
    this.reviewedAt,
    this.reviewedBy,
    this.businessName,
  });

  // Client computed metrics
  double get ctr {
    if (impressions == 0) return 0.0;
    return (clicks / impressions) * 100;
  }

  double get cpm {
    if (impressions == 0) return 0.0;
    return (budgetSpent / impressions) * 1000;
  }

  double get cpc {
    if (clicks == 0) return 0.0;
    return budgetSpent / clicks;
  }

  factory CampaignModel.fromJson(Map<String, dynamic> json) {
    // Check populating for businessName
    String? bName;
    if (json['advertiser'] is Map) {
      bName = json['advertiser']['businessName'] ?? json['advertiser']['business_name'];
    }

    return CampaignModel(
      id: json['id'] ?? '',
      advertiserId: json['advertiserId'] ?? json['advertiser_id'] ?? '',
      name: json['name'] ?? '',
      objective: json['objective'] ?? 'awareness',
      status: json['status'] ?? 'draft',
      budgetType: json['budgetType'] ?? json['budget_type'] ?? 'daily',
      budgetAmount: json['budgetAmount'] ?? json['budget_amount'] ?? 0,
      budgetSpent: json['budgetSpent'] ?? json['budget_spent'] ?? 0,
      bidStrategy: json['bidStrategy'] ?? json['bid_strategy'] ?? 'lowest_cost',
      bidAmount: json['bidAmount'] ?? json['bid_amount'],
      startDate: DateTime.parse(json['startDate'] ?? json['start_date'] ?? DateTime.now().toIso8601String()),
      endDate: json['endDate'] != null || json['end_date'] != null
          ? DateTime.parse(json['endDate'] ?? json['end_date'])
          : null,
      timezone: json['timezone'] ?? 'UTC',
      activeDays: List<int>.from(json['activeDays'] ?? json['active_days'] ?? [0, 1, 2, 3, 4, 5, 6]),
      activeHoursStart: json['activeHoursStart'] ?? json['active_hours_start'] ?? 0,
      activeHoursEnd: json['activeHoursEnd'] ?? json['active_hours_end'] ?? 23,
      targeting: Map<String, dynamic>.from(json['targeting'] ?? {}),
      placementFeed: json['placementFeed'] ?? json['placement_feed'] ?? true,
      placementReels: json['placementReels'] ?? json['placement_reels'] ?? true,
      placementStories: json['placementStories'] ?? json['placement_stories'] ?? true,
      placementExplore: json['placementExplore'] ?? json['placement_explore'] ?? false,
      impressions: json['impressions'] ?? 0,
      clicks: json['clicks'] ?? 0,
      skips: json['skips'] ?? 0,
      videoViews: json['videoViews'] ?? json['video_views'] ?? 0,
      videoCompletions: json['videoCompletions'] ?? json['video_completions'] ?? 0,
      reach: json['reach'] ?? 0,
      frequency: (json['frequency'] as num?)?.toDouble() ?? 0.0,
      websiteClicks: json['websiteClicks'] ?? json['website_clicks'] ?? 0,
      appInstalls: json['appInstalls'] ?? json['app_installs'] ?? 0,
      purchases: json['purchases'] ?? 0,
      rejectionReason: json['rejectionReason'] ?? json['rejection_reason'],
      reviewedAt: json['reviewedAt'] != null || json['reviewed_at'] != null
          ? DateTime.parse(json['reviewedAt'] ?? json['reviewed_at'])
          : null,
      reviewedBy: json['reviewedBy'] ?? json['reviewed_by'],
      businessName: bName,
    );
  }
}
